src/hitchhiker/tree/tracing_gc.cljc

```diff
@@ -0,0 +1,50 @@
+(ns hitchhiker.tree.tracing-gc
+  (:require [clojure.core.async :as async]
+            [hitchhiker.tree :as hh]
+            [hitchhiker.tree.node :as n]
+            [hitchhiker.tree.utils.async :as ha]))
+
+(defprotocol IGCScratch
+  (observe-addr! [this addr] "Marks the given addr as being currently active")
+  (observed? [this addr] "Returns true if the given addr was observed"))
+
+(defmacro do-<!
+  "Force into an async taking call.
+
+  Evaluates to <! when in an async backend.
+  Wraps form in a thread when non-async."
+  [& form]
+  (ha/if-async?
+    `(async/<! ~@form)
+    `(async/<! (async/thread ~@form))))
+
+(defn trace-gc!
+  "Does a tracing GC and frees up all unused keys.
+   This is a simple mark-sweep algorithm.
+
+   gc-scratch should be an instance of IGCScratch
+   gc-roots should be a list of the roots of currently active trees.
+   all-keys should be a core.async channel that will contain every key in storage.
+   delete-fn will be called on every key that should be deleted during the sweep phase. It is expected to return a channel that yields when the item is deleted."
+  [gc-scratch gc-roots all-keys delete-fn]
+  (let [mark-phase (async/go-loop [roots gc-roots]
```

Comment:
I would mark node ids purely in memory. Besides the slow down of writing the marked attribute to the files, storing the marks will cause problems for crashes during GC and will leave stale files behind. I also think at the moment all files that have been marked once will never be garbage collected again. This also should yield a simpler algorithm that is easier to reason about. A simple set of marked nodes should be sufficient, right?

Reply:
The design here just backs marked nodes with a konserve instance, and one can easily just back it with a memory store.

I thought having the node marking backend be configurable would be a good design; that way the user can choose whether they can do in-memory, or if they need a file-backed store.

Reply:
That makes sense, I was just wondering when you would want to store marks on disk. I can think of having a parallel GC sweep with core.async (konserve's filestore at least has non-blocking reads on the JVM) at some point. But even then you would not need to durably store the marks. Storing each mark in a separate key would definitely be very slow for IO backends.

Reply:
I am fine with merging the current version if you think it makes sense to persist the marks. I have just so far always tried to keep the abstractions minimal so that it is easier to adapt the replikativ libraries when somebody has the need.